airbyte-integrations/connectors/destination-bigquery/src/main/java/io/airbyte/integrations/destination/bigquery/typing_deduping/BigQuerySqlGenerator.java

```diff
@@ -0,0 +1,465 @@
+
+import static java.util.stream.Collectors.joining;
+
+import com.google.cloud.bigquery.StandardSQLTypeName;
+import com.google.cloud.bigquery.StandardTableDefinition;
+import com.google.cloud.bigquery.TableDefinition;
+import com.google.common.annotations.VisibleForTesting;
+import io.airbyte.integrations.base.destination.typing_deduping.AirbyteType;
+import io.airbyte.integrations.base.destination.typing_deduping.AirbyteType.AirbyteProtocolType;
+import io.airbyte.integrations.base.destination.typing_deduping.AirbyteType.Array;
+import io.airbyte.integrations.base.destination.typing_deduping.AirbyteType.OneOf;
+import io.airbyte.integrations.base.destination.typing_deduping.AirbyteType.Struct;
+import io.airbyte.integrations.base.destination.typing_deduping.AirbyteType.UnsupportedOneOf;
+import io.airbyte.integrations.base.destination.typing_deduping.AirbyteTypeUtils;
+import io.airbyte.integrations.base.destination.typing_deduping.CatalogParser.StreamConfig;
+import io.airbyte.integrations.base.destination.typing_deduping.SqlGenerator;
+import io.airbyte.integrations.destination.bigquery.BigQuerySQLNameTransformer;
+import io.airbyte.protocol.models.v0.DestinationSyncMode;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import org.apache.commons.text.StringSubstitutor;
+
+public class BigQuerySqlGenerator implements SqlGenerator<TableDefinition> {
+
+  public static final String QUOTE = "`";
+  private static final BigQuerySQLNameTransformer nameTransformer = new BigQuerySQLNameTransformer();
+
+  private final ColumnId CDC_DELETED_AT_COLUMN = buildColumnId("_ab_cdc_deleted_at");
+
+  @Override
+  public StreamId buildStreamId(final String namespace, final String name, final String rawNamespaceOverride) {
+    return new StreamId(
+        // TODO is this correct?
+        nameTransformer.getNamespace(namespace),
+        nameTransformer.convertStreamName(name),
+        nameTransformer.getNamespace(rawNamespaceOverride),
+        nameTransformer.convertStreamName(namespace + "_" + name),
+        namespace,
+        name);
+  }
+
+  @Override
+  public ColumnId buildColumnId(final String name) {
+    String quotedName = name;
+
+    // Bigquery columns are case insensitive, so do all our validation on the lowercased name
+    String canonicalized = name.toLowerCase();
+
+    // Column names aren't allowed to start with certain strings. Prepend an underscore if this happens.
+    if (canonicalized.startsWith("_table_")
+        || canonicalized.startsWith("_file_")
+        || canonicalized.startsWith("_partition_")
+        || canonicalized.startsWith("_row_timestamp_")
+        // yes, there are two underscores here. That's intentional.
+        || canonicalized.startsWith("__root__")
+        || canonicalized.startsWith("_colidentifier_")) {
+      quotedName = "_" + quotedName;
+      canonicalized = "_" + canonicalized;
+    }
+
+    // TODO this is probably wrong
```

Comment:
... 😅

Reply:
yeah... I cut a lot of corners back when I thought we didn't need to support name mangling in 1s1t >.>

we have a ticket to make sure 1s1t does this stuff correctly, right?